src/raft/raft.go

```diff
@@ -41,25 +46,319 @@ type ApplyMsg struct {
+
+	for {
+		DPrintf("👍🏻Follower state it on %v", rf.me)
+		select {
+		case <-rf.killCh:
+			close(rf.killCh)
+			DPrintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx------------------  Killed ")
+			return
+		case <-heartbeatTimeout:
+			heartbeatTimeout = randomTimeout(rf.heartbeatTimeout)
+			DPrintf("⏲ follower %v election timed out", rf.me)
+			lastContact := rf.LastContact()
+			if time.Now().Sub(lastContact) < rf.heartbeatTimeout {
+				continue
+			}
+			rf.makeCandidate()
+			return
+		case <-rf.heartBeatChan:
+			DPrintf("💖 follower %v received heartbeat", rf.me)
+			continue
+		}
+	}
+}
+
+func (rf *Raft) runCandidate() {
+	electionTimeout := randomTimeout(rf.electionTimeout)
+	rf.incTerm()
+	votes := rf.dispatchVotes()
+
+	tally := 0
+	quorum := rf.getMajority()
+
+	for {
+		select {
+		case <-rf.killCh:
+			close(rf.killCh)
+			DPrintf("xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx------------------  Killed ")
+			return
+		case <-rf.heartBeatChan:
+			DPrintf("💖 Candidate %v received heartbeat", rf.me)
+			rf.makeFollower()
+			return
+		case <-electionTimeout:
+			DPrintf("⏲ Candidate %v election timed out", rf.me)
+			return
+		case vote := <-votes:
+			if vote.Term > rf.getTerm() {
+				rf.makeFollower()
+				rf.setTerm(vote.Term)
+			}
+			if vote.VoteGranted {
+				tally++
+			}
+			if tally >= quorum {
+				DPrintf("👺 I'm the boss %v", rf.me)
+
+				rf.makeLeader()
+				return
+			}
+		}
+	}
+}
+
+func (rf *Raft) dispatchVotes() (votes chan *RequestVoteReply) {
+	// Vote for yourself
+	others := exclude(rf.peers, rf.me)
+	votes = make(chan *RequestVoteReply, len(others))
+	votes <- &RequestVoteReply{rf.getTerm(), true}
+	request := rf.makeVoteRequest()
+	for _, v := range others {
+		go func(v int) {
+			DPrintf("Requesting vote with: %+v", request)
+			reply := &RequestVoteReply{}
+			rf.sendRequestVote(v, request, reply)
+			votes <- reply
+		}(v)
+	}
+	return votes
+}
+
+func (rf *Raft) makeVoteRequest() RequestVoteArgs {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	request := RequestVoteArgs{rf.getTerm(), rf.getID()}
+
+	DPrintf("Request constructed: %+v", request)
+
+	return request
+}
+
+func (rf *Raft) makeAppendEntries(arg string) AppendEntriesArgs {
+	rf.mu.RLock()
+	defer rf.mu.RUnlock()
+	request := AppendEntriesArgs{rf.getTerm(), rf.getID()}
+	return request
+}
+
+func (rf *Raft) dispathAppendEntries() (replies chan *AppendEntriesReply) {
+	DPrintf("⏲  💖 Leader %v sent appendEntriesRPC", rf.me)
+	others := exclude(rf.peers, rf.me)
+	DPrintf("Others: %v", others)
+	replies = make(chan *AppendEntriesReply, len(others))
+	request := rf.makeAppendEntries("Ping")
+	for _, peer := range others {
+		go func(v int) {
+			DPrintf("Sending to %v Appending with: %+v", v, request)
+			reply := &AppendEntriesReply{}
+			now := time.Now()
+			ok := rf.sendAppendEntries(v, request, reply)
+			if ok {
+				diff := time.Since(now)
+				DPrintf("Request took, %+v", diff)
+			}
+			replies <- reply
+		}(peer)
+	}
+	return replies
+}
```

Comment:
Empty appendEntries RPCs are used as heartbeats.  I put in some logic to check timings of the responses.